frontend/utils/image.js

```diff
@@ -28,3 +29,99 @@ export const imageUpload = async (imageFile: File, params: ?Object): Promise<Obj
   const { download } = await sbp('chelonia/fileUpload', imageFile, { type: file.type, cipher: 'aes256gcm' }, params)
   return download
 }
+
+// Image compression
+
+export function supportsWebP (): Promise<boolean> {
+  // Uses a very small webP image to check if the browser supports 'image/webp' format.
+  // (reference: https://developers.google.com/speed/webp/faq#in_your_own_javascript)
+  const verySmallWebP = 'data:image/webp;base64,UklGRiIAAABXRUJQVlA4IBYAAAAwAQCdASoBAAEADsD+JaQAA3AAAAAA'
+  const img = new Image()
+
+  return new Promise(resolve => {
+    img.onload = () => { resolve(img.height > 0) }
+    img.onerror = (e) => { resolve(false) }
+    img.src = verySmallWebP
+  })
+}
+
+function loadImage (url): any {
+  const imgEl = new Image()
+
+  return new Promise((resolve) => {
+    imgEl.onload = () => { resolve(imgEl) }
+    imgEl.src = url
+  })
+}
+
+function generateImageBlobByCanvas ({
+  sourceImage,
+  resizingFactor,
+  quality,
+  compressToType
+}) {
+  const { naturalWidth, naturalHeight } = sourceImage
+  const canvasEl = document.createElement('canvas')
+  const c = canvasEl.getContext('2d')
+
+  canvasEl.width = naturalWidth * resizingFactor
+  canvasEl.height = naturalHeight * resizingFactor
+
+  c.drawImage(
+    sourceImage,
+    0,
+    0,
+    canvasEl.width,
+    canvasEl.height
+  )
+
+  return new Promise((resolve) => {
+    canvasEl.toBlob(blob => {
+      resolve(blob)
+    }, compressToType, quality)
+  })
+}
+
+function getResizingFactor (sourceImage) {
+  // If image's physical size is greater than the max dimension, resize the image to the max dimension.
+  const imageMaxDimension = { width: 1024, height: 768 }
```

Comment:
Let's quadruple this to 4096 x 3072

Reply:
Note that doubling it quadruples the image size. 

`2048*1536 / (1<<20) === 3`

I.e., an uncompressed image of that size would take up 3 MiB. Of course, this would generally be less with JPEG or WebP compression, but it'll come with some quality loss.

Reply:
According to Adobe, JPEG typically has a 10:1 compression ratio over BMP (BMP is essentially uncompressed):

> Nonetheless, JPEG files often achieve compression of 10:1 with no significant decrease in quality. (<https://www.adobe.com/creativecloud/file-types/image/comparison/bmp-vs-jpeg.html>)

Assuming that WebP is about 30% more efficient and a ratio of 13:1, this is about 0.70 MiB. So, a `2048*1536` will likely have reduced quality to reach the 400 KiB goal.